services/gossip/adapter/tcp/direct_transport_roundtip_metric_test.go

```diff
@@ -0,0 +1,277 @@
+			node1, node2, node3)
+
+		payloads := aMessage()
+
+		node1.listener.ExpectNotReceive()
+		node2.listener.ExpectReceive(payloads)
+		node3.listener.ExpectReceive(payloads)
+		require.NoError(t, node1.transport.Send(ctx, &adapter.TransportData{
+			SenderNodeAddress: node1.address,
+			RecipientMode:     gossipmessages.RECIPIENT_LIST_MODE_BROADCAST,
+			Payloads:          payloads,
+		}))
+
+		require.NoError(t, test.ConsistentlyVerify(test.EVENTUALLY_ADAPTER_TIMEOUT, node1.listener), "message was sent to self node")
+		require.NoError(t, test.EventuallyVerify(test.EVENTUALLY_ADAPTER_TIMEOUT, node2.listener, node3.listener), "message was not sent to target node")
+	})
+}
+
+func TestDirectTransport_UpdatesRoundtripMetric(t *testing.T) {
+	with.Concurrency(t, func(ctx context.Context, harness *with.ConcurrencyHarness) {
+		node1 := aNode(ctx, harness.Logger)
+		node2 := aNode(ctx, harness.Logger)
+		superviseAll(harness, node1, node2)
+		defer shutdownAll(ctx, node1, node2)
+
+		topology := aTopologyContaining(node1, node2)
+		node1.transport.UpdateTopology(ctx, topology)
+		node2.transport.UpdateTopology(ctx, topology)
+
+		waitForAllNodesToSatisfy(t,
+			"expected all nodes to have peers added",
+			func(node *nodeHarness) bool { return len(node.transport.outgoingConnections.activeConnections) > 0 },
+			node1, node2)
+
+		waitForAllNodesToSatisfy(t,
+			"expected all outgoing queues to become enabled after topology change",
+			func(node *nodeHarness) bool { return node.transport.allOutgoingQueuesEnabled() },
+			node1, node2)
+
+		roundtrip1to2 := node1.metricRegistry.Get(fmt.Sprintf("Gossip.OutgoingConnection.Roundtrip.%s", node2.address.String()[:6]))
+		roundtrip2to1 := node2.metricRegistry.Get(fmt.Sprintf("Gossip.OutgoingConnection.Roundtrip.%s", node1.address.String()[:6]))
+
+		for i := 1; i <= 5; i++ {
+			node1.requireSendsSuccessfullyTo(t, ctx, node2)
+			node2.requireSendsSuccessfullyTo(t, ctx, node1)
+
+			require.True(t, test.Eventually(1*time.Second, func() bool {
+				matched, err := regexp.MatchString(fmt.Sprintf("samples=%d", i), roundtrip1to2.String())
+				require.NoError(t, err)
+				return matched
+			}), "expected 1->2 roundtrip metric to update")
+
+			require.True(t, test.Eventually(1*time.Second, func() bool {
+				matched, err := regexp.MatchString(fmt.Sprintf("samples=%d", i), roundtrip2to1.String())
+				require.NoError(t, err)
+				return matched
+			}), "expected 2->1 roundtrip metric to update")
+		}
+	})
+}
+
+func replaceNodeWithSilentServer(t *testing.T, ctx context.Context, node *nodeHarness) *silentServer {
+	node.transport.GracefulShutdown(ctx)
+
+	var server *silentServer
+	var err error
+	test.Eventually(time.Second, func() bool {
+		server, err = startSilentServer(node.transport.GetServerPort())
+		return err == nil
+	})
+	require.NoError(t, err, "silent server failed to start")
+	return server
+}
+
+func TestDirectTransport_RoundtripMetricReflectsLatency(t *testing.T) {
```

Comment:
for this test, wouldn't it be easier to create a single client that talks to a server implemented in this test? why even use the big harness?

Reply:
Moved some tests to outgoing_connection_roundtrip_metric_test.go